### ISSUE TITLE
tempfile.mkdtemp() does not return absolute pathname when relative dir is specified

### ISSUE THREAD

**Issue by 013be167-c92f-46a8-afa1-7818ebd81a32** [python/cpython#51574]
BPO | [7325](https://bugs.python.org/issue7325)
--- | :---
Nosy | @pitrou, @merwok, @bitdancer, @hynek, @1st1, @iritkatriel
Files | <li>[Py7325.diff](https://bugs.python.org/file16005/Py7325.diff "Uploaded as text/plain at 2010-01-26.06:57:54 by thomas.holmes")</li><li>[Py(27)7325.diff](https://bugs.python.org/file16028/Py%2827%297325.diff "Uploaded as text/plain at 2010-01-28.04:58:14 by thomas.holmes")</li><li>[issue7325.patch](https://bugs.python.org/file19610/issue7325.patch "Uploaded as text/plain at 2010-11-15.04:04:32 by jesstess"): updated patch based on Py(27)7325.diff</li><li>[issue7325.patch](https://bugs.python.org/file19737/issue7325.patch "Uploaded as text/plain at 2010-11-21.12:00:29 by elesbom")</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2009-11-14.19:27:34.856>
labels = ['easy', 'type-bug', '3.9', '3.10', '3.11', 'library']
title = 'tempfile.mkdtemp() does not return absolute pathname when relative dir is specified'
updated_at = <Date 2021-12-06.14:33:45.013>
user = 'https://bugs.python.org/roysmith'
```

bugs.python.org fields:
```python
activity = <Date 2021-12-06.14:33:45.013>
actor = 'iritkatriel'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2009-11-14.19:27:34.856>
creator = 'roysmith'
dependencies = []
files = ['16005', '16028', '19610', '19737']
hgrepos = []
issue_num = 7325
keywords = ['patch', 'easy']
message_count = 19.0
messages = ['95256', '95257', '95307', '95308', '98316', '98317', '98342', '98405', '98406', '98457', '111828', '121214', '121734', '121764', '121906', '122038', '178291', '209838', '407816']
nosy_count = 11.0
nosy_names = ['roysmith', 'pitrou', 'dstanek', 'eric.araujo', 'r.david.murray', 'jesstess', 'thomas.holmes', 'elesbom', 'hynek', 'yselivanov', 'iritkatriel']
pr_nums = []
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue7325'
versions = ['Python 3.9', 'Python 3.10', 'Python 3.11']
```

</p></details>

### Linked PRs
* gh-94612
* gh-103844

**Comment by merwok** [comment]
@kj7rrv made a pull request but that doesn’t send a notification to people here.

The PR needs tests and docs (with versionchanged note); I think some of the original patches here contained these, so you can take inspiration from them.

**Comment by kj7rrv** [comment]
@merwok Oh, I didn't see there is already a PR for this.

**Comment by merwok** [comment]
No, I was referring to your PR!

**Comment by merwok** [comment]
The PR is ready for a final review.  I think it should not be backported, but a doc PR for older branches could be created.

**Comment by AlexWaygood** [comment]
> The PR is ready for a final review. I think it should not be backported, but a doc PR for older branches could be created.

The PR has now been merged, fixing the issue for 3.12 and newer. All that is left to do now is to file a docs PR to the 3.11 to clarify the behaviour for 3.11 and older.

@mblahay, would you be interested in doing that, maybe?

**Comment by AlexWaygood** [comment]
> I can do the doc PR if you want. Should it change "[mkdtemp()](https://docs.python.org/3/library/tempfile.html?highlight=tempfile%20mkdtemp#tempfile.mkdtemp) returns the absolute pathname of the new directory" to say "[mkdtemp()](https://docs.python.org/3/library/tempfile.html?highlight=tempfile%20mkdtemp#tempfile.mkdtemp) returns the pathname of the new directory" and add "Changed in version 3.12: the return value is always an absolute path, even if `dir` is relative"?

I realised that getting the wording right here was quite tricky, since we don't usually "document bugs", so I went ahead and made a PR myself: #103844. Reviews are welcome! :)
